File: cellPool/Docker/cellpool_procedures/20.vigCorrect.py
import skimage.io
import os
import json
import numpy as np
import sys
from basicpy import BaSiC

files = sys.argv[1:]
files = [x.strip() for x in files]
uniqueChannels = np.unique([os.path.basename(x)[7:] for x in files])

for ch in uniqueChannels:
	thisFiles = [x for x in files if ch in os.path.basename(x) and x.endswith('.tiff')]
	images = []
	for x in thisFiles:
		images.append(skimage.io.imread(x))

	images = np.array(images)
	
	b = BaSiC(get_darkfield = True, fitting_mode = 'ladmap', reweighting_tol = 0.001)
	b.fit(images)
	flatfield, darkfield = b.flatfield, b.darkfield
	imCorr = [(x - darkfield)/flatfield for x in images]
	
	for ix in range(len(thisFiles)):
		skimage.io.imsave(fname = os.path.basename(thisFiles[ix])[:-5] + \
						'.vigCorr.tiff', \
						arr = imCorr[ix], check_contrast = False)
	
	skimage.io.imsave(fname = os.path.basename(thisFiles[ix])[4:-5] + \
			'.flat-field.tiff', arr = flatfield, check_contrast = False)
	skimage.io.imsave(fname = os.path.basename(thisFiles[ix])[4:-5] + \
			'.dark-field.tiff', arr = darkfield, check_contrast = False)

File: cellPool/Docker/cellpool_procedures/20.vigCorrect.m
%This script is adapted from an example usage script available with BaSiC.
%https://github.com/marrlab/BaSiC/blob/master/examples/example_brainWSI.m

%The run command includes definition of variable 'files'

clc; close all;
addpath(genpath('BaSiC-master'));

files = string(transpose(split(files, ' ')));

uniqueChannels = unique(string(arrayfun(@(x) extractAfter(x, "-"), files)));
uniqueChannelSuffixes = unique(string(arrayfun(@(x) extractBefore(x, "."), uniqueChannels)));
thisField = string(arrayfun(@(x) extractBefore(x, "-C"), files(1)))
thisField = extractAfter(thisField, "_")
 
for i = 1:length(uniqueChannels)

	this_files = find(endsWith(files, uniqueChannels(i)));
	for j = 1:length(this_files)
		IF(:,:,j) = imread(files(this_files(j)));
	end
	
	%% estimate flatfield and darkfield
	% For fluorescence images, darkfield estimation is often necessary (set
	% 'darkfield' to be true)
	[flatfield,darkfield] = BaSiC(IF,'darkfield','true');

	%% image correction
	IF_corr = zeros(size(IF));
	for j = 1:length(this_files)
    		IF_corr(:,:,j) = (double(IF(:,:,j))-darkfield)./flatfield;
	end
	
	%% save images for next steps
	for j = 1:length(this_files)
		outName = extractBefore(files(this_files(j)), ".")
    		imwrite(uint16(IF_corr(:,:,j)), outName + '.vigCorr.tiff');
	end

	%Save flatfield and darkfield
	thisCh = char(uniqueChannelSuffixes(i))
	imwrite(flatfield, fullfile(thisField + '-' + thisCh + '.flat-field.tiff'));
	imwrite(darkfield, fullfile(thisField + '-' + thisCh + '.dark-field.tiff'));

end

